Feature: Multiple selection

fajar-apri-alaska opened the pull request:
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Resolves:** AlaskaAirlines/auro-menu#199 

## Summary:

The code diff introduces a new `multiSelect` attribute to the `AuroMenu` component, allowing for multiple options to be selected. Here's a breakdown of the functional differences:

1. **New Attribute: `multiSelect`**
   - **Definition**: Added a `multiSelect` Boolean attribute to the class, allowing the menu to support multiple selections.
   - **Usage**: 
     ```js
     this.multiSelect = undefined;
     ```

2. **Handling `value` Property**
   - **Before**: The `value` property was a single string representing the selected value.
   - **After**: The `value` property can be either a string or an array of strings when `multiSelect` is true.
     ```js
     if (!this.multiSelect) {
       this.selectByValue(this.value);
     } else {
       const values = JSON.parse(this.value);
       if (this.value && Array.isArray(values)) {
         values.forEach((val) => {
           this.selectByValue(val);
         });
       }
     }
     ```

3. **Selection State Management**
   - **New Methods**: Added methods to handle multiple selections:
     - `toggleSelectionState`: Toggles the selection state of an option.
     - `handleDeselectState`: Handles deselection of an option, updating the `value` array.
     - Modified `handleLocalSelectState` to support multiple selections by updating the `value` array.
     ```js
     toggleSelectionState(option) {
       if (option.hasAttribute('selected')) {
         this.handleDeselectState(option);
       } else {
         this.handleLocalSelectState(option);
       }
     }
     handleDeselectState(option) {
       option.removeAttribute('selected');
       option.classList.remove('active');
       option.ariaSelected = false;

       if (this.multiSelect && Array.isArray(this.value)) {
         this.value = this.value.filter((val) => val !== option.value);
         if (this.value.length === 0) {
           this.value = undefined;
         }
       }

       if (this.multiSelect && Array.isArray(this.optionSelected)) {
         this.optionSelected = this.optionSelected.filter((val) => val !== option);
         if (this.optionSelected.length === 0) {
           this.optionSelected = undefined;
         }
       }

       this.index = this.items.indexOf(option);
     }
     ```

4. **Resetting Option States**
   - **Before**: `resetOptionsStates` was always called when selecting an item.
   - **After**: `resetOptionsStates` is only called when `multiSelect` is false.
     ```js
     if (!this.multiSelect) {
       this.resetOptionsStates();
     }
     ```

5. **Event Handling**
   - Modified event handling to toggle selection state instead of always setting it:
     ```js
     if (!this.multiSelect) {
       this.handleLocalSelectState(option);
     } else {
       this.toggleSelectionState(option);
     }
     ```

The changes allow the `AuroMenu` component to handle multiple selections by using an array for the `value` property and providing methods to manage the selection and deselection of multiple options.

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [ ] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)


## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

blackfalcon commented on docs/partials/api.md:
```diff
@@ -74,6 +74,24 @@ The `auro-menu` component supports the use of the `matchWord` attribute to highl
+### Multiple
+
+`auro-menu` can have `multiple` attribute, to select multiple `auro-menuoption` elements.
+If `multiple` is enabled, `value` attribute became array value.
```
```suggestion
The `auro-menu` supports a multi-select option. To use, place the `multiSelect` attribute on the `<auro-menu>` element tag. When applied, the `value` attribute will become an Array versus String value. 
```

blackfalcon commented on src/auro-menu.js:
```diff
@@ -15,7 +15,8 @@ import "mark.js/dist/mark.min";
  * @attr {Boolean} disabled - When true, the entire menu and all options are disabled;
  * @attr {Boolean} noCheckmark - When true, selected option will not show the checkmark.
- * @attr {String} value - Value selected for the menu.
+ * @attr {Boolean} multiple - When true, selected option can be multiple.
```
```suggestion
 * @attr {Boolean} multiSelect - When true, the selected option can be multiple.
```

blackfalcon commented on src/auro-menu.js:
```diff
@@ -15,7 +15,8 @@ import "mark.js/dist/mark.min";
  * @attr {Boolean} noCheckmark - When true, selected option will not show the checkmark.
- * @attr {String} value - Value selected for the menu.
+ * @attr {Boolean} multiple - When true, selected option can be multiple.
+ * @attr {String} value - Value selected for the menu, can be String or Array if multiple is true.
```
```suggestion
 * @attr {String} value - Value selected for the menu, can be String or Array if `multiSelect` is true.
```

blackfalcon commented on src/auro-menu.js:
```diff
@@ -41,6 +42,7 @@ export class AuroMenu extends LitElement {
     this.noCheckmark = false;
     this.ready = false;
     this.optionActive = undefined;
+    this.multiple = undefined;
```
```suggestion
    this.multiSelect = undefined;
```

blackfalcon commented on src/auro-menu.js:
```diff
@@ -62,6 +64,7 @@ export class AuroMenu extends LitElement {
       optionActive:   { type: Object },
       matchWord:      { type: String },
       ready:          { type: Boolean },
+      multiple:       { type: Boolean },
```
```suggestion
      multiSelect:       { type: Boolean },
```

blackfalcon commented on src/auro-menu.js:
```diff
@@ -100,10 +103,26 @@ export class AuroMenu extends LitElement {
+
     if (changedProperties.has('value')) {
-      this.selectByValue(this.value);
+      if (!this.multiple) {
```
```suggestion
      if (!this.multiSelect) {
```

blackfalcon commented on src/auro-menu.js:
```diff
@@ -164,6 +183,50 @@ export class AuroMenu extends LitElement {
+    option.ariaSelected = false;
+
+    // remove from this.value if there is any, set to undefined if became empty array
+    if (this.multiple && Array.isArray(this.value)) {
```
```suggestion
    if (this.multiSelect && Array.isArray(this.value)) {
```

blackfalcon commented on src/auro-menu.js:
```diff
@@ -164,6 +183,50 @@ export class AuroMenu extends LitElement {
+    }
+
+    // do the same for selected option DOM
+    if (this.multiple && Array.isArray(this.optionSelected)) {
```
```suggestion
    if (this.multiSelect && Array.isArray(this.optionSelected)) {
```

blackfalcon commented on src/auro-menu.js:
```diff
@@ -174,8 +237,29 @@ export class AuroMenu extends LitElement {
 
-    this.value = option.value;
-    this.optionSelected = option;
+    if (this.multiple) {
```
```suggestion
    if (this.multiSelect) {
```

blackfalcon commented on src/auro-menu.js:
```diff
@@ -208,7 +292,9 @@ export class AuroMenu extends LitElement {
 
     if (this.items[this.index] && !this.items[this.index].hasAttribute('disabled')) {
-      this.resetOptionsStates();
+      if (!this.multiple) {
```
```suggestion
      if (!this.multiSelect) {
```

blackfalcon commented on test/auro-menu.test.js:
```diff
@@ -264,8 +264,68 @@ describe('auro-menu', () => {
+  })
 });
 
+it('test select/deselect options in multiple', async () => {
```
```suggestion
it('test select/deselect options in multiSelect', async () => {
```

blackfalcon commented on test/auro-menu.test.js:
```diff
@@ -264,8 +264,68 @@ describe('auro-menu', () => {
+  expect(menuEl.children[0].hasAttribute('selected')).to.equal(false);
+})
+
+it('test changing values after render in multiple', async () => {
```
```suggestion
it('test changing values after render in multiSelect', async () => {
```

blackfalcon commented on test/auro-menu.test.js:
```diff
@@ -359,3 +419,18 @@ async function emptyItemsFixture() {
+async function multipleMenuFixture() {
+  return await fixture(html`
+    <div>
+      <auro-menu multiple aria-label="test">
```
```suggestion
      <auro-menu multiSelect aria-label="test">
```

blackfalcon commented on test/auro-menu.test.js:
```diff
@@ -359,3 +419,18 @@ async function emptyItemsFixture() {
+      </auro-menu>
+    </div>
+  `);
+}
```
Please add an empty return.

fajar-apri-alaska replied:
Sorry, what do you mean by empty return for this one? it is currently returns the fixture.

blackfalcon commented:
While this is functionally living up to expectations, we have a pretty big a11y/UX issue. Please see the attached video. Now that we have this option to multi-select more than one menu item, using the keyboard arrow keys to move up and down the list to unselect any options is almost impossible. 

See in the video that the DOM is triggering active state, but the UI is NOT reflecting this feedback making this almost unseable to assistive device use. 

@leeejune we will need some assistance with this. 

![May-14-2024 12-43-19](https://github.com/AlaskaAirlines/auro-menu/assets/181089/1599316b-ef23-469d-9ae7-9d88bfd1d18d)

leeejune commented:
@blackfalcon I see a few suggestions to address this:
1. We use checkboxes for multi select menu items. I don't know how the behavior is inherited, but I see the "press space to select" behavior being inherited. That should solve this issue of selecting with up/down arrows.
2. We implement the "press enter to select" behavior like how it is shown here: https://semantic-ui.com/modules/dropdown.html

blackfalcon commented:
> @blackfalcon I see a few suggestions to address this:
> 
> 1. We use checkboxes for multi select menu items. I don't know how the behavior is inherited, but I see the "press space to select" behavior being inherited. That should solve this issue of selecting with up/down arrows.
> 2. We implement the "press enter to select" behavior like how it is shown here: https://semantic-ui.com/modules/dropdown.html

Some of the recommendations you are suggesting will be implemented with the new multi-select combobox element where the selected options will appear in a list above the dropdown. But that's not the issue I am bringing up here. 

I am electing to move this to a new issue where we can discuss a keyboard interaction solution. 

https://github.com/AlaskaAirlines/auro-formkit/issues/149

blackfalcon committed:
feat(multi-select): add support for selection option #199

The core of this commit is updating the ability of the element to know
the difference between the default single-select and the multi-select
option. With the multi-select options selected, the value of the menu
selections will be kept in an array versus a single string.

Changes to be committed:
modified:   src/auro-menu.js
modified:   src/style-base.scss
modified:   src/style-menuoption.scss

blackfalcon committed:
docs(multi-select): add new UI demo examples

Changes to be committed:
new file:   apiExamples/multiple.html
modified:   docs/partials/api.md

blackfalcon committed:
docs: remove unused content

A reference string in the page was removed as it's only template
content.

Changes to be committed:
modified:   docs/partials/index.md
modified:   demo/index.md

blackfalcon committed:
test(multi-select): update test cases for new feature

Changes to be committed:
modified:   test/auro-menu.test.js

blackfalcon committed:
build: update build scripts

This commit updates the pakcage.json build scripts to be compliant with
the decisions made in this discussion:

https://github.com/orgs/AlaskaAirlines/discussions/523

This commit also updates the `assets` object to reference the correct
api.md file for post-back to the repo.

Changes to be committed:
modified:   .github/workflows/testPublish.yml
modified:   package.json

blackfalcon committed:
chore: update dependencies

Changes to be committed:
modified:   package-lock.json
modified:   package.json

blackfalcon reviewed (dismissed):
I updated this commit and PR to fit an evolving specification the team is going for. Please let me know if you have any questions. 

Also, please look into this weird warning coming from the test logs on a build

https://github.com/AlaskaAirlines/auro-menu/actions/runs/9166453632/job/25201867531#step:6:25

fajar-apri-alaska commented:
> I updated this commit and PR to fit an evolving specification the team is going for. Please let me know if you have any questions.
> 
> Also, please look into this weird warning coming from the test logs on a build
> 
> https://github.com/AlaskaAirlines/auro-menu/actions/runs/9166453632/job/25201867531#step:6:25

I am aware of this as well.
Since there is a [part of the code](https://github.com/AlaskaAirlines/auro-menu/blob/eef024c2a613343f6a95830bb7aca1910191f29c/src/auro-menu.js#L111-L120) where it reads the `value` attributes (which it can be an array). So I am putting that inside try catch function to avoid function break.

jason-capsule42 commented:
@sourcery-ai review

chrisfalaska commented:
@blackfalcon I intend on picking this up and integrating it into `auro-formkit`.

Commit:
fix: json parse error, merge fixes

Commit:
fix: optimize multi-select function

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/AlaskaAirlines/auro-menu?pullRequest=202) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/AlaskaAirlines/auro-menu?pullRequest=202) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: blackfalcon<br/>:x: chrisfalaska<br/><hr/>**chrisfalaska** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/AlaskaAirlines/auro-menu?pullRequest=202) it.</sub>

Commit:
chore: satisfy linting

Commit:
chore: update docs

chrisfalaska commented:
This work has been moved to `auro-formkit` and will be completed in https://github.com/AlaskaAirlines/auro-formkit/issues/123
